Add Crane 2.0 enhancement proposals (#288)
konveyor/enhancements: enhancements/crane-2.0/validate-api-compatibility/README.md

A CI/CD pipeline runs `crane validate` after `crane apply`. The non-zero exit code from `crane validate` halts the pipeline before any cluster modification occurs.

#### Story 3: Post-Transform Validation Confirms Fixes

After running `crane transform` with plugins that remap deprecated APIs, the operator runs the full pipeline and validates the final output. All entries report OK; exit code 0. The operator proceeds to `kubectl apply` with confidence.

#### Story 4: CRD-Backed Custom Resources

An application uses a CRD-backed type `widgets.example.com/v1`. The CRD is not installed on the target cluster. `crane validate` reports the incompatibility. The operator installs the CRD on the target before applying.

#### Story 5: Air-Gapped / Disconnected Environment (Phase 2 -- Offline Mode)

The target cluster is on a restricted network. A team member with cluster access captures the API surface with `kubectl api-resources -o json`. The migration team validates without cluster access using `crane validate --api-resources api-resources.json`. The matching logic, report format, and exit codes are identical to live mode.

#### Story 6: Pre-Provisioning Check (Phase 2 -- Offline Mode)

> The target cluster doesn't exist yet, but the team knows it will be OCP 4.14. Someone captures the API surface from a reference cluster and uses it for validation.

Reviewer: I think we should consider the possibility of packaging API resources bundles of well known Kubernetes distributions and versions. In your example, if the user doesn't have access to an OCP 4.14 cluster, but knows that the target is effectively OCP 4.14, it should be possible to just specify that with a flag for simplicity sake.

Reviewer 2: Good idea for a future enhancement. 

For now, anyone with access to a reference cluster of the target version can capture and share the API surface file. 

Bundling well-known distributions is interesting but i think it carries a maintenance burden. Also what all distributions we will be maintaining? Only for OCP?

Author: @stillalearner , This is about OCP to OCP for coming version.

Author: In a discussion we had it was agreed to implement "Enable the CLI to fetch relevant metadata automatically based on a provided "OCP version" input" in a later version.

Author: Created https://github.com/migtools/crane/issues/387 to track this feature for a later version.

The command enforces `cobra.NoArgs` -- positional arguments are rejected to prevent typos like `crane validate ./manifests` from silently validating the default directory.

**Mutual exclusion (Phase 2):**

When `--api-resources` is NOT set, existing behavior is unchanged -- `crane validate` uses the default kubeconfig as it does today. No breaking changes.

#### Pipeline Position

Validate operates on the **final rendered manifests** from `crane apply`'s output directory, not the raw export or intermediate transform output. This is deliberate:

* **Export** produces raw resources from the source cluster
* **Transform** runs plugins that may modify apiVersions, add/remove fields
* **Apply** renders final YAML via `kubectl kustomize` into the output directory
* **Validate** checks those final manifests against the target cluster's API surface (live or offline)
* **`kubectl apply`** deploys to the target cluster

Validating after apply (rendering) ensures that transform plugins' changes are reflected in the check. Validating raw exports would produce false positives for resources that transforms will fix.

> 

Reviewer 3: Suggestion: For better clarity, could we add this section on resource filtering ? 

**Resource Filtering:**  The runtime-generated resources like Pods, ReplicaSets, Endpoints etc. are  filtered out by `crane transform`. The default KubernetesPlugin (along with other transform plugins if configured) identifies
  and excludes controller-managed resources from the final `kustomization.yaml` , keeping only declarative
  resources that users should apply to the target cluster.

  Here's an example of resource count through the pipeline 
  * **Export stage:** 13 resources (includes Pod, ReplicaSet, Endpoints, EndpointSlice, default ServiceAccount, kube-root-ca.crt ConfigMap)
  * **Transform stage:** Filters to 7 declarative resources (removes the 6 runtime-generated resources above)
  * **Validate stage:** Reports 7 resources scanned

Reviewer 2: yes, we can have something like this

This enhancement adds a new **`crane validate`** command -- a read-only preflight check that scans the final rendered manifests from `crane apply`'s output directory, queries the **target** cluster's discovery API, and reports which apiVersion+kind combinations are not available on the target under strict group/version matching. When an incompatible resource's kind exists under a different apiVersion on the target, validate suggests the alternative. This brings Crane toward feature parity with mig-controller's GVK compatibility warnings, giving operators early visibility into API skew between clusters.

The enhancement is delivered in phases:

* **Phase 1 (implemented):** Live-cluster validation -- `crane validate` queries the target cluster's discovery API directly via kubeconfig.

> * **Phase 2 (planned):** Offline/disconnected validation -- `crane validate --api-resources api-resources.json` accepts the JSON output of `kubectl api-resources -o json` for air-gapped environments, pre-provisioning checks, and CI/CD pipelines without cluster access.

Reviewer 2: ```suggestion
* **Phase 2 (planned):** Offline/disconnected validation -- ` crane validate --api-resources api-surface.json accepts a captured API surface JSON file (produced by scripts/capture-api-surface.sh)` for air-gapped environments, pre-provisioning checks, and CI/CD pipelines without cluster access.
```

### Goals

* **Early detection:** Surface all API incompatibilities between final rendered manifests and the target cluster *before* `kubectl apply`, in a single report.
* **CI-friendly:** Provide deterministic exit codes so `crane validate` can gate a migration pipeline -- non-zero exit when incompatible GVKs are found.
* **Read-only:** Validate never modifies manifests or cluster state. It is a pure reporting command.
* **Low privilege:** Validate only requires basic cluster authentication -- it uses the Kubernetes discovery API (`/api`, `/apis`) which is accessible to all authenticated users via the default `system:discovery` ClusterRole. No special RBAC is needed.
* **Pipeline composability:** Validate fits naturally into the export -> transform -> apply -> validate flow, reading from `crane apply`'s output directory without requiring changes to other commands.
* **Suggestion engine:** When an incompatible resource's kind is available under a different apiVersion on the target, suggest the alternative to guide the user.

> * **Offline/disconnected support (Phase 2):** Support validation against air-gapped clusters by accepting `kubectl api-resources -o json` output as input, using the same matching logic as live mode.

Reviewer 2: ```suggestion
* **Offline/disconnected support (Phase 2):** Support validation against air-gapped clusters by aaccepting an API surface JSON file as input, using the same matching logic as live mode.
```

#### Story 2: CI Pipeline Gates on Validation

A CI/CD pipeline runs `crane validate` after `crane apply`. The non-zero exit code from `crane validate` halts the pipeline before any cluster modification occurs.

#### Story 3: Post-Transform Validation Confirms Fixes

After running `crane transform` with plugins that remap deprecated APIs, the operator runs the full pipeline and validates the final output. All entries report OK; exit code 0. The operator proceeds to `kubectl apply` with confidence.

#### Story 4: CRD-Backed Custom Resources

An application uses a CRD-backed type `widgets.example.com/v1`. The CRD is not installed on the target cluster. `crane validate` reports the incompatibility. The operator installs the CRD on the target before applying.

#### Story 5: Air-Gapped / Disconnected Environment (Phase 2 -- Offline Mode)

> The target cluster is on a restricted network. A team member with cluster access captures the API surface with `kubectl api-resources -o json`. The migration team validates without cluster access using `crane validate --api-resources api-resources.json`. The matching logic, report format, and exit codes are identical to live mode.

Reviewer 2: ```suggestion
The target cluster is on a restricted network. A team member with cluster access  captures the API surface using scripts/capture-api-surface.sh . The migration team validates without cluster access using `crane validate --api-resources api-surface.json`. The matching logic, report format, and exit codes are identical to live mode.
```

#### Story 6: Pre-Provisioning Check (Phase 2 -- Offline Mode)

The target cluster doesn't exist yet, but the team knows it will be OCP 4.14. Someone captures the API surface from a reference cluster and uses it for validation.

#### Story 7: CI/CD Pipeline Without Cluster Access (Phase 2 -- Offline Mode)

> For multi-cluster targeting, the pipeline runs validate once per cluster using the corresponding `api-resources.json` file.

Reviewer 2: ```suggestion
For multi-cluster targeting, the pipeline runs validate once per cluster using the corresponding API surface JSON file.
```

### Implementation Details/Notes/Constraints

#### Command Structure

New command at `cmd/validate/validate.go`, following the existing Cobra pattern (self-contained package, options struct with Complete/Validate/Run, PreRun with viper binding).

**Flags (Phase 1 -- implemented):**

| Flag | Description |
|------|-------------|
| `--input-dir` / `-i` | Root directory of final rendered manifests to scan (default: `"output"`, matching crane apply's `--output-dir` default) |
| `--validate-dir` | Directory where validation report and failure artifacts are saved (default: `"validate"`) |
| `--kubeconfig` | Path to target cluster kubeconfig |
| `--context` | Kubeconfig context for target cluster |
| `--output` / `-o` | Report file format: `json` (default) or `yaml`. Controls the format of the persisted report file. Table output is always printed to the terminal. |

**Additional flag (Phase 2 -- planned):**

| Flag | Description |
|------|-------------|

> | `--api-resources` | Path to `kubectl api-resources -o json` output file. Enables offline validation. Mutually exclusive with `--kubeconfig`/`--context`. |

Reviewer 2: ```suggestion
| `--api-resources` | Path to API surface JSON file for offline validation. Mutually exclusive with `--kubeconfig`/`--context`. |
```

* Decode each document individually -- one bad document logs a warning and continues to the next, without aborting remaining documents in the file
* Extract `apiVersion`, `kind`, `metadata.namespace` from each document
* Skip documents missing `apiVersion` or `kind` (e.g., empty documents, comments-only documents)
* Skip the `failures/` subdirectory to avoid re-scanning previous validation output
* Collect distinct (group, version, kind, namespace) tuples, tracking source files for each
* Deduplicate by GVK+namespace and sort results by group/version/kind/namespace

Input filenames are sanitized when writing failure artifacts to prevent path traversal from malicious manifest content.

#### Target Discovery

**Live mode (Phase 1):** Use client-go discovery client:

* Call `ServerGroupsAndResources()` on the target cluster
* Handle partial discovery failures gracefully (continue with available groups, log warning)
* Call `discoveryClient.Invalidate()` before querying to ensure fresh results (no stale cache)
* Build a two-level lookup map: groupVersion -> kind -> APIResource

> **Offline mode (Phase 2):** Parse `kubectl api-resources -o json` output:

Reviewer 2: ```suggestion
**Offline mode (Phase 2):**  Parse the API surface JSON file (format: {"apiResourceLists": [<APIResourceList>, ...]} containing native metav1.APIResourceList objects per group-version):
```

> * Deserialize the native Kubernetes `APIResourceList` JSON with `json.Unmarshal`

Reviewer 2: ```suggestion
* Deserialize with json.Unmarshal into the same Go types the discovery client return
```

* Build the same two-level lookup map as live mode

> * Zero custom parsing -- the JSON output uses the same Go types (`metav1.APIResource`) the discovery client returns internally

Reviewer 2: ```suggestion
```

> The JSON output from `kubectl api-resources -o json` provides everything needed:

Reviewer 2: ```suggestion
```

> | JSON field | Maps to | Used for |

Reviewer 2: ```suggestion
```

|------------|---------|----------|
| `group` + `version` | `groupVersion` (e.g., `apps/v1`; core resources omit `group`, so groupVersion = `v1`) | GVK matching |
| `kind` | `kind` | GVK matching |
| `name` | `resourcePlural` | Report output |

> | `namespaced` | `namespaced` | Future: scope validation |

Reviewer 2: ```suggestion

```

**Permissions (live mode):** Only requires the Kubernetes discovery API (`/api`, `/apis`), which is accessible to all authenticated users via the default `system:discovery` ClusterRoleBinding -> `system:authenticated` group. No namespace-level RBAC or resource-read permissions are needed.

**Permissions (offline mode):** No cluster access needed from the migration workstation. The person capturing the API resources file needs only basic authentication on the target cluster.

> 

Reviewer 2: ```suggestion

### Capture Script

  The `scripts/capture-api-surface.sh` script captures the full API surface of a target cluster into a single JSON file for offline validation. It queries every served group-version via the raw API endpoints and produces
   the `{"apiResourceLists": [...]}` format that `crane validate --api-resources` expects.

  <details>
  <summary>View script</summary>

  ```bash
  #!/bin/bash
  # capture-api-surface.sh
  # Run this on a machine with kubectl access to the target cluster.
  # It captures the full API surface (all served versions, not just preferred).
  # Pass the output file to: crane validate --api-resources api-surface.json
  #
  # Usage:
  #   bash capture-api-surface.sh --context <context-name> -o <output-file>
  #   bash capture-api-surface.sh --kubeconfig <path> -o <output-file>
  #   bash capture-api-surface.sh --kubeconfig <path> --context <context-name> -o <output-file>
  #
  # Examples:
  #   bash capture-api-surface.sh --context my-target-cluster -o api-surface.json
  #   bash capture-api-surface.sh --kubeconfig /path/to/kubeconfig -o api-surface.json

  set -euo pipefail

  CONTEXT=""
  KUBECONFIG_PATH=""
  OUTPUT=""

  while [[ $# -gt 0 ]]; do
    case "$1" in
      --context)    CONTEXT="$2"; shift 2 ;;
      --kubeconfig) KUBECONFIG_PATH="$2"; shift 2 ;;
      -o)           OUTPUT="$2"; shift 2 ;;
      *)            echo "Unknown flag: $1"; exit 1 ;;
    esac
  done

  if [ -z "$OUTPUT" ]; then
    echo "Usage: bash capture-api-surface.sh [--context <name>] [--kubeconfig <path>] -o <output-file>"
    exit 1
  fi

  # Build kubectl flags
  KUBECTL_FLAGS=""
  [ -n "$CONTEXT" ] && KUBECTL_FLAGS="$KUBECTL_FLAGS --context=$CONTEXT"
  [ -n "$KUBECONFIG_PATH" ] && KUBECTL_FLAGS="$KUBECTL_FLAGS --kubeconfig=$KUBECONFIG_PATH"

  echo "Capturing API surface..."
  [ -n "$CONTEXT" ] && echo "  Context: $CONTEXT"
  [ -n "$KUBECONFIG_PATH" ] && echo "  Kubeconfig: $KUBECONFIG_PATH"

  # Step 1: Get all API versions served by the cluster
  API_VERSIONS=$(kubectl api-versions $KUBECTL_FLAGS)

  # Step 2: For each API version, fetch the list of resources it serves
  echo '{"apiResourceLists":[' > "$OUTPUT"
  FIRST=true

  for GV in $API_VERSIONS; do
    # Core API (v1) lives at /api/v1, everything else at /apis/<group>/<version>
    if [ "$GV" = "v1" ]; then
      ENDPOINT="/api/v1"
    else
      ENDPOINT="/apis/$GV"
    fi

    RESOURCES=$(kubectl get --raw "$ENDPOINT" $KUBECTL_FLAGS 2>/dev/null) || continue

    if [ "$FIRST" = true ]; then
      FIRST=false
    else
      echo "," >> "$OUTPUT"
    fi
    echo "$RESOURCES" >> "$OUTPUT"
  done

  echo ']}' >> "$OUTPUT"

  COUNT=$(echo "$API_VERSIONS" | wc -l | tr -d ' ')
  echo "Done. Captured $COUNT API versions to $OUTPUT"

  Usage:

  # On a machine with target cluster access:
  bash scripts/capture-api-surface.sh --context my-target-cluster -o api-surface.json

  # On the migration workstation (no cluster access needed):
  crane validate --api-resources api-surface.json
  
```

**Target cluster access (live mode):** Validate requires read access to the target cluster's discovery API. This is a low-privilege operation -- discovery is available to all authenticated users via the default `system:discovery` ClusterRole. No special RBAC setup is needed. The kubeconfig must be handled securely.

**No write operations:** Validate is strictly read-only on the cluster. It only writes to the local filesystem (report and failure artifacts under `--validate-dir`).

**Path traversal prevention:** Failure artifact filenames are derived from manifest content (kind, apiVersion, namespace). All components are sanitized through `safeFilePart()` which strips path separators and non-alphanumeric characters, preventing directory escape.

**False positives with CRDs:** If a CRD is installed on the target but its API group hasn't been aggregated yet (e.g., during CRD installation), validate may report a false incompatibility. Documentation should note this timing consideration.

**Stale discovery cache (live mode):** The command calls `discoveryClient.Invalidate()` before querying to ensure fresh results from the target cluster's API server.

> **Stale data (offline mode):** The `kubectl api-resources -o json` file is a point-in-time snapshot. Users should re-capture after cluster upgrades or CRD changes. The report includes the source file path and mode for traceability.

Reviewer 2: ```suggestion
**Stale data (offline mode):** The API surface JSON file is a point-in-time snapshot. Users should re-capture after cluster upgrades or CRD changes.
```

* **Regression:** Existing export/transform/apply tests unaffected (validate is additive)

**Phase 2 (planned):**

* **Unit tests:** JSON parsing -- valid output, empty resources, malformed JSON, core resources (no group), non-core resources (with group), duplicate kinds across groups
* **Unit tests:** Offline matching -- same scenarios as live matcher tests but with a parsed index from JSON instead of a discovery client
* **Unit tests:** Mutual exclusion -- `--api-resources` with `--context` or `--kubeconfig` returns error
* **Integration tests:** Capture `kubectl api-resources -o json` from a test cluster, run `crane validate --api-resources` with it -- ensure results match live validate against the same cluster
* **E2E tests:** Cross-platform scenarios (EKS api-resources validated against OCP manifests)
* **Regression:** Live-mode validate behavior unchanged

### Offline Mode: Edge Cases (Phase 2)

| Concern | Handling |
|---------|----------|
| **Stale data** | Document that the file is a point-in-time snapshot. Users should re-capture after cluster upgrades or CRD changes. |

> | **CRDs installed after capture** | Same as above -- re-run the kubectl command. |

Reviewer 2: ```suggestion
| **CRDs installed after capture** | Same as above -- re-run the capture script. |
```

| **Empty resources array** | Error: "api-resources file contains no resources" |
| **Malformed JSON** | Error from `json.Unmarshal` with clear message. |
| **Missing `kind` field at top level** | Warn if `kind` is not `APIResourceList`, but still attempt to parse. |

> | **Core resources (no `group` field)** | `group` is omitted for core API resources. Parser uses `version` alone as `groupVersion` (e.g., `"v1"`). |

Reviewer 2: ```suggestion
| **Core resources (no `group` field)** | Core API resources use groupVersion: "v1" (no group prefix). The parser handles this naturally since groupVersion is provided directly. |
```

| **Duplicate resources across groups** | Multiple entries with the same kind but different group/version are valid (e.g., `Event` in `v1` and `events.k8s.io/v1`). Index handles this naturally since it's keyed by `groupVersion`. |
| **Verbs / RBAC** | Out of scope. `verbs` field is present in JSON but ignored. |
| **Multi-cluster** | `--api-resources` takes one file. For N clusters, run validate N times with the corresponding file. |

> | **kubectl version compatibility** | `kubectl api-resources -o json` is available since K8s 1.11+. The `APIResourceList` schema is stable. |

Reviewer 2: ```suggestion
| **kubectl version compatibility** | `The API surface JSON format wraps native metav1.APIResourceList objects. The capture script uses standard discovery APIs available since K8s 1.11+. |
```

## Implementation History

* **#185:** Exploration of mig-controller's resource versioning logic -- design input for this enhancement. *(Open)*
* **#230:** Feature story for this enhancement. *(Open)*
* **#302:** Phase 1 implementation PR -- crane validate command with live-cluster validation, 23 unit tests. *(Merged)*
* **#319:** Phase 2 feature request -- offline/disconnected cluster support. *(Open)*

## Drawbacks

* **Additional step in the pipeline:** Users must remember to run `crane validate` -- it is not automatically invoked by `crane apply`. This is deliberate (composability), but means users can still skip it.
* **Strict matching may over-report:** Some resources may work on the target under a different version of the same group (e.g., v1beta1 -> v1), but strict matching flags them as incompatible. The suggestion engine mitigates this by showing available alternatives.
* **Discovery limitations (live mode):** The Kubernetes discovery API may not surface all available API versions in all configurations (e.g., aggregated API servers with intermittent availability). The command handles partial discovery failures gracefully.

> * **Stale data (offline mode):** The api-resources JSON file is a point-in-time snapshot that may drift from the actual cluster state. Documentation should emphasize re-capturing after changes.

Reviewer 2: ```suggestion
* **Stale data (offline mode):** The API surface JSON file is a point-in-time snapshot that may drift from the actual cluster state. Documentation should emphasize re-capturing after changes.
```

## Alternatives

1. **Integrate validation into crane apply as a preflight step:** Rejected because it couples a read-only check to a write command, violating Crane's Unix-philosophy composability. Users may want to validate without being ready to apply, or validate in a different environment than where apply runs.
2. **Validate at export time against both source and target:** Rejected because export is source-only by design, and requiring target access during export adds a dependency that doesn't exist today. Validate as a separate command keeps concerns cleanly separated.
3. **Transform plugin that rewrites apiVersions:** Complementary, not a replacement. A transform plugin can fix known version mappings, but the user still needs a way to discover *which* mappings are needed. Validate provides that discovery.
4. **Static compatibility database:** Ship a hardcoded map of "version X removed in Kubernetes Y." Rejected because it cannot cover CRDs, custom API servers, or OpenShift-specific APIs, and requires constant maintenance.

> 5. **Custom offline format (Phase 2 alternative):** Define a crane-specific YAML format for target API surface. Rejected in favor of `kubectl api-resources -o json` because it requires zero custom parsing, uses native Kubernetes types, and needs no crane tooling on the target cluster side.

Reviewer 2: ```suggestion
5. **Custom offline format (Phase 2 alternative):** Define a crane-specific YAML format for target API surface. Rejected in favor of the API surface JSON format which wraps native Kubernetes APIResourceList types and only requires a lightweight capture script (scripts/capture-api-surface.sh) on the target cluster side.
```
